### DISCUSSION TITLE
Reorganize Standard Library

### DISCUSSION THREAD

**Discussion by magnus-madsen** [flix/flix#6546] [Ideas]
I would like to open a discussion on reorganization of the standard library. While the current structure has served us well, I think we cannot keep putting everything into the root module. We should try to break it down.

I do not plan to make any changes before "qualified uses" are supported (i.e. https://github.com/flix/flix/issues/6515). Thus if module `X` is moved to `Y.Z.W` a qualified use of `Y.Z.W.X` will restore current functionality. Moreover, we will designate a collection of default qualified uses. For example, `List` should always be available imho.

The current library is here: https://api.flix.dev/

We should look for inspiration among Haskell, Rust, Scala cats, etc.

With that, here is a proposal to get started. 

```
Algebra
  Applicative
  Collectable (?)
  CommutativeGroup
  CommutativeMonoid
  CommutativeSemiGroup, 
  Filterable
  Foldable
  Functor
  Group
  Identity
  Monad
  MonadZero
  MonadZip
  Monoid
  Reducible
  SemiGroup
  Traversable
  UnorderedFoldable
  Witherable

Core
  BigInt
  BigDecimal
  Bool
  Char
  CodePoint (?)
  Float32
  Float64
  Int8
  Int16
  Int32
  Int64
  Ref
  String

Core.Order
  Comparison
  Down
  Eq
  Hash
  JoinLattice
  LowerBound
  MeetLattice
  Order
  PartialOrder
  UpperBound

Core.Logic.Fixpoint
  ...

Core.Ops
  {Add, Div, Mul, Neg, Sub}

Concurrency
  Channel

Data
  Array
  Chain
  DelayList
  DelayMap
  List
  Map
  MultiMap
  MutDeque
  MutList
  MutMap
  MutQueue
  MutSet
  Nec
  Option
  RedBlackTree
  Result
  Set
  Validation
  Vector

Lang
  Adaptor
  Assert
  Boxable
  Boxed
  Closeable
  FromString
  Iterable (?)
  Iterator (?)
  Object
  Proxy
  Reflect
  ToJava
  ToString

System
  Console
  Environment
  File
  Files
  IOError
  Random
  Reader
  Thread

Util
  Benchmark
  GetOpt
  Graph
  Regex
  StringBuilder
```

Default Qualified Uses:
```
use Data.{List, Map, Option, Result, Set}
```

We should try to give each root module some meaning. Here is an attempt:

- `Algebra` contains mathematical abstractions (e.g. `Monoid`).
- `Core` modules somehow essential to the language.
- `Core.Ops` type classes for `+`, `-` and so forth.
- `Data` contains collections and collection-like modules.
- `System` contains things for the platform and things likely to have the `IO` effect.
- `Util` contains "mini-library" modules.

We should also come up with some rules like `Util` can depend on everything, but `Core` should not depend on `Util`.

**Comment by mlutze** [comment]
We might want to consider prefixing the whole standard library with `Flix`, and making a namespacing standard, like mlutze/flix-json should start all its namespaces as `FlixJson` or something like that.

**Comment by JonathanStarup** [comment]
> `Data` contains collections and collection-like modules.

Sounds like it should be called `Collections`

**Comment by stephentetley** [comment]
I think Matthew's suggestion to have a top level namespace is a good one - for me it's something that Java got right and Haskell got wrong.

Maybe `Text` would be a useful major group with `Text.toString` and `Text.StringBuilder` rather than `Lang.ToString` and `Util.StringBuilder`.

**Reply by magnus-madsen** [reply]
I think part of splitting String from StringBuilder is that it would be useful if we could only load String without having to load StringBuilder. And in general, if something is in e.g. Util (or whatever) we would only have to load it if it used.

**Comment by magnus-madsen** [comment]
A benefit of using modules is also that dependencies between modules would become much more clear.

**Comment by magnus-madsen** [comment]
@mlutze @JonathanStarup Do either of you have a screenshot of our sorting exercise? I cannot find my pictures.

**Reply by mlutze** [reply]
![20231106_161040.jpg](https://github.com/flix/flix/assets/19153692/85f63ab9-203c-488d-90f8-418b1d81e8fd)

![20231107_140841.jpg](https://github.com/flix/flix/assets/19153692/a55e1eba-5693-4042-b002-7eaaad72faf5)

![20231106_161122.jpg](https://github.com/flix/flix/assets/19153692/32692f22-2a05-4d21-bd25-556876b53b4a)

![20231107_140836.jpg](https://github.com/flix/flix/assets/19153692/b15ebd36-ba9e-444f-847f-30c1c9f5aa3e)

![20231106_161030.jpg](https://github.com/flix/flix/assets/19153692/ebed5665-4328-46b2-b863-137a9ce45b34)

**Comment by magnus-madsen** [comment]
I would like to get started on this slowly. I plan to make a series of small issues to refactor very gradually. 

(Just because I make an issue does not mean its settled. We can still discuss in that issue.)

@stephentetley Would you be interested in helping out? I ask because you invested a lot in the standard library.

**Comment by stephentetley** [comment]
Thanks Magnus, yes I'll be interested in helping out. I'd got a bit stuck with effects not being ready and I had to get up to speed with Python for work, but now effects are coming online I'll be prioritizing Flix again.

**Comment by LionelMeli** [comment]
About adding Range into the library. Could we add as usual in Range module or in Data.Range one?

**Comment by magnus-madsen** [comment]
I think Data.Range
